### PR TITLE
Enhancement reporting benchmark results

### PR DESCRIPTION
**Pull request by AronisAt79**
This PR aims to enhance the reporting functionality for github actions-triggered circuit benchmarks. Results are saved in postgresql and accessible via grafana

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

- New python module introduced for results processing and saving.
- getSysStats.sh will now only display the test timers
- New action step: Calculate Benchmark Result prints a grafana url linking to detailed test results and downloadable prover log file
- New sysstat collection script (sadf.sh) now handles statistics for test that span beyond 00.00 hrs
- execBench.sh : Added missing bench tests (bytecode, exp, copy, pi)

### CODE REVIEW

**Comment by AronisAt79** [comment]
Sample result
https://grafana.zkevm-testnet.org/d/vofy8DAVz/circuit-benchmarks?orgId=1&var-test_id=Super-19-Benchmark-PR1166--1676306463&from=1676303840589&to=1676306473805

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_main.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+                    usage = 'python3 reporting_main.py 13 EVM 19',
+                    description = 'Writes circuit benchmark results to postgresql, uploads logfiles to s3 bucket',
+                    )
+    parser.add_argument('pr') 
+    parser.add_argument('circuit')
+    parser.add_argument('degree')
+    parser.add_argument('test_id')
+    args = parser.parse_args()
+    pr, circuit, degree, test_id = (args.pr, args.circuit, args.degree, args.test_id)
+    # circuit = circuit.split(":")[1].replace('"', '')
```
Comment (ChihChengLiang) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_main.py:21]:
```suggestion
```

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                'cpu_all_Max'           : sysstat['cpu_all_Max'],
+                'cpu_count'             : sysstat['cpu_count'],
+                'sysstats_url'          : f'{env["grafana_dashboard_prefix"]}{test_id}',
+                'logsurl'               : f'{env["s3endpoint"]}{test_id}.tar.gz'
+            }
+
+    except Exception as e:
+        print(e)
+
+    # print(r['sysstats_url'])
```
Comment (ChihChengLiang) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:68]:
```suggestion
```

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    engine   = create_engine(f'postgresql://{user}:{password}@{host}:5432/{database}')
+
+    return engine
+
+def prepare_result_dataframe(test_result, sysstat,env, test_id):
+    '''
+    prepares postgres data (in the form of dataframe) for table circuit_benchmarks
+    '''
+    try:
+        # test_id = f'{test_result["circuit"]}-{test_result["degree"]}-Benchmark-PR{test_result["pull_request"]}_{str(time.time_ns())[13:]}'
```
Comment (ChihChengLiang) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:27]:
```suggestion
```

**Review by ChihChengLiang — APPROVED** [review]
LGTM. Added some nitpicks to remove deadcode comments.

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_main.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    engine = rmod.pgsql_engine(env['db'])
+    data.to_sql(table,engine,if_exists='append')
+    ms = rmod.write_mem_time(engine,memstats, test_id)
+    cs = rmod.write_cpuall_time(engine,cpustats, test_id)
+
+    url = f'{env["grafana_dashboard_prefix"]}{test_id}'.replace(" ", "")
+    print(f'Test Result: {url}')
+
+if __name__ == '__main__':
+    main()
```
Comment (CPerezz) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_main.py:34]: Add a new line at the end of the file :)

**Comment by AronisAt79** [comment]
> LGTM. Added some nitpicks to remove deadcode comments.

Thanks @ChihChengLiang . Comments gone.

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    '''
+    accepts ram stats raw data  and returns a dataframe for further processing
+    '''
+    statsdf = df[['timestamp', 'kbmemused']]
+    # statsdf = statsdf.set_index('timestamp')
+    statsdf['kbmemused']   = pd.to_numeric(statsdf['kbmemused'])
+    statsdf['utilizationgb']   = statsdf['kbmemused'].apply(lambda x: round(x/float(1000000),2))
+    statsdf = statsdf[['timestamp','utilizationgb']]
+    max_ram = statsdf['utilizationgb'].max()
+    return statsdf, max_ram
```
Comment (CPerezz) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:212]: ditto

**Review by CPerezz — COMMENTED** [review]
I wonder if we can add some comments on the README regarding how to access/check the benchmark results. 

Are they publicly avaliable?

**Comment by AronisAt79** [comment]
> I wonder if we can add some comments on the README regarding how to access/check the benchmark results.
> 
> Are they publicly avaliable?

Thanks @CPerezz. Yes they are public. Will add a comment.

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                'setup_gen'             : test_result['setup_gen'],
+                'proof_gen'             : test_result['proof_gen'],
+                'proof_ver'             : test_result['proof_ver'],
+                'max_ram'               : sysstat['max_ram'],
+                'cpu_all_Average'       : sysstat['cpu_all_Average'],
+                'cpu_all_Max'           : sysstat['cpu_all_Max'],
+                'cpu_count'             : sysstat['cpu_count'],
+                'sysstats_url'          : f'{env["grafana_dashboard_prefix"]}{test_id}',
+                'logsurl'               : f'{env["s3endpoint"]}{test_id}.tar.gz'
+            }
```
Comment (ChihChengLiang) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:62]: Both `r`s look the same in both branches.

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    except Exception as e:
+        print(e)
+        return None
+
+def process_cpustats(statsdf):
+    '''
+    accepts cpu stats raw data from csv and returns a dataframe for further processing
+    '''
+    statsdf = statsdf[['timestamp', '%idle']]
+    # statsdf = statsdf.set_index('timestamp')
```
Comment (ChihChengLiang) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:193]:
```suggestion
```

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    cpu_all_Average = statsdf['utilizationall'].mean()
+    return statsdf, cpu_all_Max,cpu_all_Average
+
+
+def process_memstats(df):
+    '''
+    accepts ram stats raw data  and returns a dataframe for further processing
+    '''
+    statsdf = df[['timestamp', 'kbmemused']]
+    # statsdf = statsdf.set_index('timestamp')
```
Comment (ChihChengLiang) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:207]:
```suggestion
```

**Review by ChihChengLiang — APPROVED** [review]
Found some more nitpicks. Feel free to ignore them

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py (hunk 5 above)
Comment (AronisAt79) [reply .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:212]: ends with new line but only shown in edit mode :)

**Comment by AronisAt79** [comment]
> I wonder if we can add some comments on the README regarding how to access/check the benchmark results.
> 
> Are they publicly avaliable?

@CPerezz i added section ## GH Actions Benchmark Results in README

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py (hunk 6 above)
Comment (AronisAt79) [reply .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py]: Thanks @ChihChengLiang , the conditional was obviously redundant. Has been removed with[14c169c](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1166/commits/14c169cd5dad5e411430ec1209eb970a7ec1ba7d)
which also fixes a typo and allows parser to read the full prover log

**Comment by AronisAt79** [comment]
@ChihChengLiang @CPerezz i realized that the prover logs were not consistently formatted in circuit-benchmarks. I submitted https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1207 to make them uniform. I will retest this branch once #1207 is merged and then request for your re-review, if that is ok.

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    if running != '0':
+        r = [i.split(":")[1].split(".")[0].replace(" ","") for i in logdata if 'test result' in i][0]
+        if r == 'ok':
+            result = 'PASSED'
+            try:
+                sg = ''.join(g[0] for g in itertools.groupby([i for i in logdata if 'End' in i and 'Setup generation' in i ][0])).split('.', 1)[-1]
+            except:
+                sg = 'None'
+            try:
+                pg = ''.join(g[0] for g in itertools.groupby([i for i in logdata if 'End' in i and 'Circuit Proof generation' in i ][0])).split('.', 1)[-1]
```
Comment (ChihChengLiang) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:108]: Sorry, I think I misunderstood #1207. 
You were looking for a match of "Circuit Proof generation." But before #1207, some benchmarks have messages like "Bytecode Proof generation with {} rows," which doesn't match.

So in your original proposed change of #1207, you changed  "Bytecode Proof generation with {} rows" to "Bytecode Circuit Proof generation with {} rows." But I mistook you were looking for a match of "Bytecode Circuit" and gave review feedback that adds constant `BENCHMARK_ID="Bytecode Circuit"`. After applying that change, it did not make the parsing clear.

This is the rust code post #1207

```rust
// circuit-benchmarks/src/bytecode_circuit.rs

//Unique string used by bench results module for parsing the result
const BENCHMARK_ID: &str = "Bytecode Circuit";
let proof_message = format!("{} Proof generation with degree = {}", BENCHMARK_ID, degree);
```
And now, in the Python script, we are looking for a substring match for 'Circuit Proof generation' from the output `proof_message`, which is hard to reason.

---
Here's my proposal.

1. We add common constants in bench_params for parsing. 

```rust
// bench_params.rs
// These prefixes are designed to be parsed by the benchmark server.
const SETUP_PREFIX: &str = "[Setup generation]";
const PROOF_PREFIX: &str = "[Proof generation]";
```

2. In each benchmark code, we add the constants to the messages

```rust
let proof_message = format!("{} {} with degree = {}", PROOF_PREFIX, BENCHMARK_ID, degree);
```

3. In the reporting_modules.py, we add constants too.

```python
# We search for the prefix in the logs of the Rust code.
PROOF_PREFIX = '[Proof generation]'
pg = ''.join(g[0] for g in itertools.groupby([i for i in logdata if 'End' in i and PROOF_PREFIX in i ][0])).split('.', 1)[-1]
```

We can apply all the above changes in this PR, so the parsing logic is visible in one place.

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+    
+    return cpustats, memstats, sysstat
+
+def log_processor(pull_request,circuit, degree):
+    '''
+    Exports test metadata and result metrics from prover logfile
+    '''
+    logfile = [i for i in os.listdir('/home/ubuntu/') if 'proverlog' in i][0]
+    f = open(f'/home/ubuntu/{logfile}', 'r')
+    logdata = f.read()
```
Comment (ChihChengLiang) [inline .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:96]: Note for reviewers: 
The logdata looks like

```
running 1 test
Start:   Setup generation with degree = 18
End:     Setup generation with degree = 18 .........................................16.129s
Start:   EVM circuit Proof generation with degree = 18
test evm_circuit::evm_circ_benches::bench_evm_circuit_prover has been running for over 60 seconds
End:     EVM circuit Proof generation with degree = 18 .............................561.062s
Start:   EVM circuit Proof verification
End:     EVM circuit Proof verification ............................................104.904ms
test evm_circuit::evm_circ_benches::bench_evm_circuit_prover ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 8 filtered out; finished in 587.50s
```

**Review by ChihChengLiang — COMMENTED** [review]
I propose we change how we parse the log

File: .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py (hunk 9 above)
Comment (AronisAt79) [reply .github/proverCiScripts/benchmarks_result_reporting/reporting_modules.py:108]: > You were looking for a match of "Circuit Proof generation." But before #1207, some benchmarks have messages like "Bytecode Proof generation with {} rows," which doesn't match.

My initial intention was to simply make sure that we have common constant patterns in the logs, that the parser would always match. I think your first proposal served that purpose fine, but yes, this approach is cleaner and more consistent.

**Review by ChihChengLiang — APPROVED** [review]
LGTM! Thanks for the fix. 
Can you run a `make fmt` to fix some formatting issues?

**Pull request merged**